### PR TITLE
[DefiniteInitialization] Check whether globals captured by top-level defer statements are initialized.

### PR DESCRIPTION
**Pull request by ravikandhadai**
<rdar://30720636>

### CODE REVIEW

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

**Review by rjmccall — APPROVED** [review]
LGTM.

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

**Comment by ravikandhadai** [comment]
Pushed a new commit that resolves merge conflicts

File: lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp [hunk 1]
```diff
@@ -1243,10 +1244,13 @@ void LifetimeChecker::handleEscapeUse(const DIMemoryUse &Use) {
 
   Diag<StringRef, bool> DiagMessage;
   if (isa<MarkFunctionEscapeInst>(Inst)) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp]: I am fine if you want to do that in a subsequent commit.

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

File: lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ravikandhadai) [reply lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp]: Okay. I have added some explanation to this piece of code. My exact change here is to just introduce a new `else if` condition to the logic in the  "MarkFunctionEscapeInst" branch, which is something I understand well. The rest of the logic existed before. However, I have documented my overall understanding of the code here.

**Review by gottesmm — COMMENTED** [review]
Thats better than what we have here. I think the prose can be cleaned up, but I think you are leaving this better than you found it, so I am fine with it.

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge

**Pull request merged**